### ISSUE TITLE
Speed setting is not always correct

### ISSUE THREAD

**Issue by Danstiv** [espeak-ng/espeak-ng#1461]
espeak sets the speed incorrectly under some conditions.

#### 1
Voice files use the speed parameter, which allows to set the base speed for the voice, or something like that.

This parameter does not work if espeak build without sonic.

#### 2
If espeak build without sonic, you can set any high speed for it, for example 10000.

In this case, espeak will generate speech at the maximum allowed speed, but the resulting data will be slightly shorter than when using espeak with sonic at speed 449.

In other words, it is possible to achieve faster speech without distortion from espeak without sonic than from espeak with sonic.

I think this is incorrect behavior, the problem is probably somewhere in the SetSpeed function, but my knowledge of C and understanding of what is happening there is not enough to determine the problem more precisely.

**Comment by Danstiv** [comment]
I realize that the issue doesn't look very convincing without examples, so I've gathered some data to confirm the problem.

For testing, I used espeak with sonic and without sonic built from commit 819cab294e278ffe86be1177d1467d6c255d9895.

The test document was readme.md from the nvda repository (commit 9a30254dc17b7329f0e353296b7c19bc45c04e48).

| Test | With sonic | Variant | Espeak speed | Voice speed | Result length |
| --- | --- | --- | --- | --- | --- |
| 1 | yes | none | 449 | not set | 499.207 |
| 2 | yes | max | 449 | not set | 499.207 |
| 3 | yes | none | 449 | 200000 | 498.791 |
| 4 | yes | max | 449 | 200000 | 499.207 |
| 5 | no | none | 449 | not set | 493.924 |
| 6 | no | none | 449 | 200000 | 493.508 |
| 7 | no | max | 449 | 200000 | 493.924 |
| 8 | no | none | 200000 | not set | 493.508 |
| 9 | no | none | 200000 | 200000 | 493.508 |
| 10 | no | max | 449 | not set | 493.924| 
| 11 | no | max | 200000 | not set | 493.508 |
| 12 | no | max | 200000 | 200000 | 493.508 |

**Comment by jaacoppi** [comment]
> I think this is incorrect behavior, the problem is probably somewhere in the SetSpeed function

Correct. "speed" and other voice parameters are handled in voices.c. In short, it sets voice->speed_percent. That's used in SetSpeed(). There's two versions of the function, one for Sonic and one without. It seems they duplicate a lot of code. This is very bad programming practise. I'll try to refactor them so we can spot the difference.

**Comment by Danstiv** [comment]
> This parameter does not work if espeak build without sonic.

I'm sorry, I was wrong.

This parameter does not work if espeak build WITH sonic.

In any case, when looking for a problem, it is better to use the table from the comment above.

**Comment by jaacoppi** [comment]
See https://github.com/espeak-ng/espeak-ng/pull/1465.

I have refactored SetSpeed() to be more understandable.

espeakRATE_MAXIMUM is 450. There's a clue for the value 449.

**Comment by Danstiv** [comment]
Nothing changed.

Or was it just refactoring?

| Test | With sonic | Variant | Espeak speed | Voice speed | Result length | new result length |
| --- | --- | --- | --- | --- | --- | --- |
| 1 | yes | none | 449 | not set | 499.207 | 499.207 |
| 2 | yes | max | 449 | not set | 499.207 | 499.207 |
| 3 | yes | none | 449 | 200000 | 498.791 | 498.791 |
| 4 | yes | max | 449 | 200000 | 499.207 | 499.207 |
| 5 | no | none | 449 | not set | 493.924 | 493.924 |
| 6 | no | none | 449 | 200000 | 493.508 | 493.508 |
| 7 | no | max | 449 | 200000 | 493.924 | 493.924 |
| 8 | no | none | 200000 | not set | 493.508 | 493.508 |
| 9 | no | none | 200000 | 200000 | 493.508 | 493.508 |
| 10 | no | max | 449 | not set | 493.924 | 493.924 |
| 11 | no | max | 200000 | not set | 493.508 | 493.508 |
| 12 | no | max | 200000 | 200000 | 493.508 | 493.508 |

**Comment by Danstiv** [comment]
> espeakRATE_MAXIMUM is 450. There's a clue for the value 449.

Sound crackling appears at speed 450 with sonic, it seems that this is also a bug.

**Comment by jaacoppi** [comment]
Yeah all I did was refactor so the problem would be easier to debug spot.

**Comment by Danstiv** [comment]
Voice speed is applied regardless of the presence of Sonic (when building from master), sonic triggering threshold was fixed in #2165, so all problems described are solved.
